lib/charms/layer/containerd.py

```diff
@@ -1,4 +1,68 @@
 from charmhelpers.core import hookenv, unitdata
+from charmhelpers.core.hookenv import resource_get
+from functools import lru_cache
+import os
+from pathlib import Path
+from subprocess import check_call, check_output, CalledProcessError
+from typing import Union
+
+
+@lru_cache()
+def arch():
+    """Determine current machine's arch."""
+    return check_output(["dpkg", "--print-architecture"]).decode().strip()
+
+
+class ResourceFailure(Exception):
+    """Custom exception to raise when resource isn't viable."""
+
+    pass
+
+
+def unpack_containerd_resource() -> Union[None, Path]:
+    """Unpack containerd resource and provide pathh to parent directory."""
+    try:
+        archive = resource_get("containerd")
+    except Exception:
+        return ResourceFailure("Error fetching the containerd resource.")
+
+    if not archive:
+        return ResourceFailure("Missing containerd resource.")
+
+    charm_dir = os.getenv("CHARM_DIR")
+    unpack_path = Path(charm_dir, "resources", "containerd")
+    return _unpack_archive(archive, unpack_path)
+
+
+def _unpack_archive(archive, unpack_path):
+    unpack_path.mkdir(exist_ok=True, parents=True)
+    archive = Path(archive)
+    filesize = archive.stat().st_size
+    if filesize == 0:
+        return None
+    if filesize < 10000000:
+        return ResourceFailure("Incomplete containerd resource")
+    check_call(["tar", "xfz", archive, "-C", unpack_path])
+    return _collect_resource_bins(unpack_path)
+
+
+def _collect_resource_bins(unpack_path):
+    for arch_based in unpack_path.glob("./*.tar.gz"):
+        if arch() in arch_based.name:
+            unpack_path = unpack_path / arch()
+            return _unpack_archive(arch_based, unpack_path)
```

Comment:
check to see if the original archive contains `tar.gz` files within it -- if so it's likely they are the arch based resources